### DISCUSSION TITLE
version `GFORTRAN_10' not found

### DISCUSSION THREAD

**Discussion by chenyubi14** [nakib/elphbolt#75] [General]
Hi,

I compiled the packages successfully. However, when I tried to do the test example by running the binary `elphbolt.x`, I ran into this issue
> elphbolt.x: /lib64/libgfortran.so.5: version `GFORTRAN_10' not found (required by elphbolt.x)

I searched this issue online, found two methods in this [website](https://stackoverflow.com/questions/9628273/libgfortran-version-gfortran-1-4-not-found) , but both of them don't work.

Here is the setup of my system. I am working on a cluster (pod on cnsi.ucsb.edu) where I don't have the privilege to edit /lib64/. The installed `libgfortran.so` in the cluster has versions lower than `GFORTRAN_10`. I found that the package `OpenCoarrays/ ` has `libgfortran.so` that supports `GFORTRAN_10` in two paths: OpenCoarrays/prerequisites/installations/gnu/12.1.0/lib64/libgfortran.so and OpenCoarrays/prerequisites/builds/gcc-12.1.0/x86_64-pc-linux-gnu/libgfortran/.libs/libgfortran.so.

1. The first method is to link libgfortran.so to the `libgfortran.so` in `OpenCoarrays/ `. It failed because I cannot edit /lib64/
2. The second method is to use LD_PRELOAD. I put LD_PRELOAD=${HOME}/OpenCoarrays/prerequisites/builds/gcc-12.1.0/x86_64-pc-linux-gnu/libgfortran/.libs/libgfortran.so in my `~/.bashrc `file. I faced this error

> ERROR: ld.so: object '~/OpenCoarrays/prerequisites/builds/gcc-12.1.0/x86_64-pc-linux-gnu/libgfortran/.libs/libgfortran.so' from LD_PRELOAD cannot be preloaded: ignored.

Both methods failed, but I think there might be a way to change the path to `libgfortran.so.5` in `elphbolt.x`. I tried to search where this path is defined in the source files, but I failed. I wonder whether you could point out a way to solve this issue.

THANKS!

**Comment by nakib** [comment]
Hi chenyubi14,

Thanks for using `elphbolt` and opening this issue.

I think the following combination worked for me for building `OpenCoarrays 2.9.2`: `gnu_gcc/9.2.0`, `openmpi/4.0.5gcc9.2.0`, and `cmake/3.18.4`. You can specify the versions of these dependencies when you build `OpenCoarrays`. The documentation of `elphbolt` gives a couple of examples. Let me know if this fixes your issue.

Have you managed to successfully build `elphbolt` on a desktop or a laptop?

Best,
Nakib

**Comment by chenyubi14** [comment]
Hi Nakib,

Thanks a lot for your response!

I don't really think the suggestions could resolve my issue. In any case, I have successfully compiled `elphbolt` on my ubuntu system, and it works smoothly. Thus, it is not urgent for me to use elphbolt in the cluster.

If you want to hear my long-winded problem, here it is. My problem cannot be solved by building `OpenCoarrays` as your suggested, because I have already successfully compiled `OpenCoarrays`. My cluster does not have the new versions of gcc as required by `OpenCoarrays`, so I can not load modules like what you did. As a result, `OpenCoarrays` installed gcc, mpich, etc, in its subfolder with the new versions. Therefore, I have `GFORTRAN_10` supported in a subfolder of `OpenCoarrays/`, but `elphbot.x` cannot use a `libgfortran.so.5` that supports `GFORTRAN_10`.

I think there is a way to link `elphbot.x` to the fortran in the subfolders of `OpenCoarrays/`, so this is why I am posting this issue. If my descriptions are not clear enough, you may also ignore it, and I can continue using `elphbolt` with my ubuntu.

THANKS!

**Comment by nakib** [comment]
Good to know that you were able to build the program on your Ubuntu machine. For production level calculations, however, you will probably have to move on to a cluster at some point. To that end, I can suggest that you ask your cluster maintainer if they can create a `singularity` container from a `dockerfile` (https://docs.sylabs.io/guides/2.6/user-guide/singularity_and_docker.html). Here's a `dockerfile` which I have tested on my laptop: https://github.com/nakib/elphbolt/blob/develop/Dockerfile. This should be a completely portable solution. Let me know if this is a possibility for your cluster. 

Another option is a `spack` (https://spack.readthedocs.io/en/latest/) package. Let me know if you are interested and I will share with you an example, but I am not an expert on it by any means. :)

Best,
Nakib

**Comment by chenyubi14** [comment]
Hi Nakib,

I have tried building  docker image by the dockerfile, and install it by singularity. However, I seem to have the preload issues, so we gave up on this method. 

With the help of the cluster managers, we finally have elphbolt.x working properly. The final working method is by using "module load intel/20;module load openmpi/2.1.0" and "include msi_intel.make" on Pod.

Thanks,
Yubi

**Comment by nakib** [comment]
Dear Yubi

Good to know that you managed to get the code running on your cluster. Have fun!

Best
Nakib

**Comment by chenyubi14** [comment]
I need to point out one more issue, and my final solution.

If I use "msi_intel.make", there are other problems. One major problem is that no matter how many cores I assigned, like `cafrun -n 2 elphbolt.x`, it will always use 40 cores. In addition, the jobs may not properly finish.

The reason is that "msi_intel.make" uses `FC=ifort`, but `FC=caf` is needed to resolve the problems.

However, `FC=caf` like in "lp_gfortran.make" and "thickpad_gcc.make" will have the very first issue I mentioned. 
`elphbolt.x: /lib64/libgfortran.so.5: version GFORTRAN_10 not found (required by elphbolt.x)`

My final solution is add `export LD_LIBRARY_PATH=/<path-to-OpenCoarrays>/prerequisites/installations/gcc/<release-tag>/lib64:$LD_LIBRARY_PATH` to `~/.bashrc`. (I got the hint from this website https://github.com/sourceryinstitute/gcc/ )
It will have `FC=caf` working perfectly, and the number of cores exactly follows my command `cafrun -n 2 elphbolt.x`
Note OpenCoarrays should be compiled by `./install.sh`

**Reply by nakib** [reply]
Hi Yubi,

For an intel build, there is no need to use `cafrun`. The `caf` and `cafrun` commands are for use with `gcc+opencoarrays`. For an intel build, you can try something like this in your run script:
```
export FOR_COARRAY_NUM_IMAGES=2
<path to your elphbolt executable>/elphbolt.x
```

Best,
Nakib
